### PR TITLE
[WIP][RFC][AC SMP merge] add outbound rpc for SMP

### PR DESCRIPTION
**Pull request by christinacelee**
## Summary

This is second PR for the AC-SMP refactor/merge project - first PR in stack is https://github.com/libra/libra/pull/1870
- adding parallel processes that handle periodic broadcasting of batches of transactions via RPC

TODO: 
- potentially add more tests (especially for killing worker thread for disconnected peer)
- add logging/error-handling (esp. in RPC response proto structs)
- `shared_mempool` and some tests use hard-coded values. Replace them with actual values from config
- remove code that these changes deprecate

## Test Plan

Refactored `network/src/validator_network/test.rs` and `shared_mempool_test`, which pass

## Related PRs
First PR in stack: https://github.com/libra/libra/pull/1870

### CODE REVIEW

**Comment by christinacelee** [comment]
for now don't accept/land - just review please! Land should wait until `futures` version is only 0.3 in `mempool` in master

File: config/src/config/network_config.rs [hunk 1]
```diff
@@ -81,6 +82,50 @@ impl Default for NetworkConfig {
...
+    pub fn default_validator_network_with_peers(
+        nodes_in_network_count: u32,
+    ) -> HashMap<PeerId, Self> {
+        let mut peers_network_config: HashMap<PeerId, Self> = HashMap::new();
+        let mut peers_keypairs: HashMap<PeerId, NetworkKeyPairs> = HashMap::new();
+        let mut all_peers: Vec<PeerId> = Vec::new();
+
+        // create peers
+        let mut rng = StdRng::from_seed(TEST_SEED);
+        for _n in 0..nodes_in_network_count {
```
Comment (phoenix-antigravity) [inline config/src/config/network_config.rs]: nit: just `_` instead of `_n`

File: config/src/config/network_config.rs [hunk 2]
```diff
@@ -81,6 +82,50 @@ impl Default for NetworkConfig {
 }
 
 impl NetworkConfig {
+    /// Creates a validator network of with `nodes_in_network_count` validator nodes
+    pub fn default_validator_network_with_peers(
```
Comment (phoenix-antigravity) [inline config/src/config/network_config.rs]: looks like this method is only used in tests. Is it possible to define it there?

File: mempool/src/shared_mempool.rs [hunk 3]
```diff
@@ -379,27 +541,22 @@ where
         .expect("[shared mempool] failed to create runtime");
     let executor = runtime.handle();
 
-    let peer_info = Arc::new(Mutex::new(PeerInfo::new()));
+    let validator_peers = if let Some(v) = &config.validator_network {
+        v.network_peers.peers.iter().map(|(key, _)| *key).collect()
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool.rs]: nit: use match syntax instead

File: mempool/src/shared_mempool.rs [hunk 4]
```diff
@@ -31,28 +42,40 @@ use tokio::{
 };
 use vm_validator::vm_validator::{get_account_state, TransactionValidation};
 
+const WORKER_CHANNEL_BUFFER_SIZE: usize = 5;
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool.rs]: what's a reason to have it set to 5? How smp will behave if network keeps blipping and worker can't process messages in time?

**Comment by davidiw** [comment]
@christinacelee are you still working on this PR?

**Comment by christinacelee** [comment]
this is tabled for now - I can close for now and reopen when relevant

**Pull request closed without merging**